pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py

```diff
@@ -0,0 +1,443 @@
+            if d.sem.acquire(timeout=timeout):
+                d.sem.release()
+            else:
+                raise TimeoutError('Wait case to finish timeout')
+
+    def _start_sub_case_thread(
+        self,
+        dev_res: MultiDevResource,
+        case: UnittestMenuCase,
+        sub_case_index: int,
+        retry_count: int = DEFAULT_RETRY_COUNT,
+    ) -> None:
+        """
+        Start the thread monitoring on the corresponding dut of the sub-case
+        """
+        # Allocate the kwargs that pass to '_run'
+        _kwargs = {}
+        _kwargs['dut'] = dev_res.dut
+        _kwargs['dev_res'] = dev_res
+        _kwargs['case'] = case
+        _kwargs['sub_case_index'] = sub_case_index
+        _kwargs['retry_count'] = retry_count
+
+        # Create the thread of the sub-case
+        dev_res.thread = Thread(target=self._run, kwargs=_kwargs, daemon=True)
+        dev_res.thread.start()
+        # Thread starts, acquire the semaphore to block '_wait_multi_dev_case_finish'
+        dev_res.sem.acquire()
+
+    def _run(self, **kwargs) -> None:  # type: ignore
+        """
+        The thread target function
+        Will run for each case on each dut
+
+        Call the wrapped function to trigger the case
+        Then keep listening on the dut for the signal
+
+            - If the dut send a signal, it will be put into others' recv_sig
+            - If the dut waits for a signal, it block and keep polling for the recv_sig until get the signal it requires
+            - If the dut finished running the case, it will quite the loop and terminate the thread
+        """
+        signal_pattern_list = [
+            self.UNITY_SEND_SIGNAL_REGEX,  # The dut send a signal
+            self.UNITY_WAIT_SIGNAL_REGEX,  # The dut is blocked and waiting for a signal
+            unity.UNITY_SUMMARY_LINE_REGEX,  # Means the case finished
+        ]
+        dut = kwargs['dut']
+        dev_res = kwargs['dev_res']
+        case = kwargs['case']
+        sub_case_index = kwargs['sub_case_index']
+        retry_count = case.attributes.get('retry', kwargs['retry_count'])
+        # Start the case
+        dut.expect(self.ready_pattern_list)
+        # Retry at defined number of times if not write successfully
+        for retry in range(retry_count):
+            dut.write(str(case.index))
+            try:
+                dut.expect_exact(case.name, timeout=1)
+                break
+            except TIMEOUT as e:
+                if retry >= retry_count - 1:
+                    dev_res.sem.release()
+                    raise e
+
+        dut.write(str(sub_case_index))
+
+        # Wait for the specific patterns, only exist when the sub-case finished
+        while True:
+            pat = dut.expect(signal_pattern_list, timeout=60)
+            if pat is not None:
+                match_str = pat.group().decode('utf-8')
+
+                # Send a signal
+                if match_str.find(self.SEND_SIGNAL_PREFIX) >= 0:
```

Comment:
As a noob, I learnt that this might be simplified to `if self.SEND_SIGNAL_PREFIX in match_str:`. Same as following `elif`